Mathlib/Computability/MyhillNerode.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2024 Google. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Wong
+-/
+import Mathlib.Computability.DFA
+import Mathlib.Data.Set.Finite
+
+/-!
+# Myhill–Nerode theorem
+
+This file proves the Myhill–Nerode theorem via left quotients.
+
+Given a language `L` and a word `x`, the *left quotient* is the set of suffixes `y` such that
+`x ++ y` is in `L`. The *Myhill–Nerode theorem* shows that each left quotient, in fact, corresponds
+to the state of an automaton that matches `L`, and that `L` is regular if and only if there are
+finitely many such states.
+
+## Implementation notes
+
+Since mathlib doesn't have an `IsRegular` definition yet, we define regularity directly as the
+existence of a DFA.
+
+## References
+
+* <https://en.wikipedia.org/wiki/Syntactic_monoid#Myhill%E2%80%93Nerode_theorem>
+-/
+
+universe u v
+variable {α : Type u} {σ : Type v} (L : Language α)
+
+namespace Language
+
+/-- The *left quotient* of `x` is the set of suffixes `y` such that `x ++ y` is in `L`. -/
+def leftQuotient (x : List α) : Language α := { y | x ++ y ∈ L }
+
+@[simp]
+theorem leftQuotient_nil : L.leftQuotient [] = L := rfl
+
+@[simp]
+theorem leftQuotient_append (x y : List α) :
+    L.leftQuotient (x ++ y) = (L.leftQuotient x).leftQuotient y := by
```

Comment:
I doubt this is a good idea as a simp lemma

Reply:
I don't have a strong opinion on this. Removing `@[simp]` here does mean we have to add it manually to the proof of `toDFA_accepts` below.

Reply:
I guess if we want to make this simp, then `evalFrom_of_append` must be simp too.

Reply:
(I removed the `@[simp]` for now.)